features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/presequite/SessionReplayRequirement.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.presequite
+
+import com.datadog.android.api.InternalLogger
+import okhttp3.internal.toImmutableMap
+
+internal class SessionReplayRequirement private constructor(
+    private val checkers: List<SystemRequirementChecker>,
+    private val systemRequirementConfiguration: SystemRequirementConfiguration
+) {
+
+    fun runIfMeetRequirements(
+        featureName: String,
+        internalLogger: InternalLogger,
+        runnable: () -> Unit
+    ) {
+        if (checkAllMinimumRequirements()) {
+            runnable()
+        } else {
+            internalLogger.log(
+                level = InternalLogger.Level.INFO,
+                listOf(InternalLogger.Target.TELEMETRY, InternalLogger.Target.USER),
+                messageBuilder = {
+                    "$featureName is disabled because the system " +
+                        "doesn't meet the minimum SDK requirements"
+                },
+                onlyOnce = true,
+                additionalProperties = getCheckerReport()
+            )
+        }
+    }
+
+    private fun checkAllMinimumRequirements(): Boolean {
+        if (checkers.isEmpty()) {
+            return true
+        }
+        return checkers.all {
+            it.checkMinimumRequirement(systemRequirementConfiguration)
+        }
+    }
+
+    private fun getCheckerReport(): Map<String, Any?> {
+        val map = mutableMapOf<String, Any?>()
+        checkers.forEach { checker ->
+            map[checker.name()] = checker.checkedValue()
+        }
+        return map.toImmutableMap()
```

Comment:
you can simply call `toMap` here instead if you want non-mutable instance or just `return map`, because the return type specified is read only.

Reply:
updated